File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/DummyW3CActionAdapter.java
package io.appium.espressoserver.lib.helpers.w3c.adapter;

import android.graphics.Point;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
import io.appium.espressoserver.lib.handlers.exceptions.NoSuchElementException;
import io.appium.espressoserver.lib.handlers.exceptions.NotYetImplementedException;
import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException;
import io.appium.espressoserver.lib.helpers.Logger;
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.W3CKeyEvent;
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType;
import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState;
import io.appium.espressoserver.lib.helpers.w3c.state.PointerInputState;

public class DummyW3CActionAdapter extends BaseW3CActionAdapter {

    // Keep a log of pointer move events so the values can be checked in the unit tests
    private List<PointerMoveEvent> pointerMoveEvents = new ArrayList<>();

    public static class PointerMoveEvent {
        public String sourceId;
        public PointerType pointerType;
        public long currentX;
        public long currentY;
        public long x;
        public long y;
        public Set<Integer> buttons;
        public KeyInputState globalKeyInputState;
    }

    private class DummyLogger implements Logger {
        public void error(Object... messages) {
            // No-op
        }
        public void error(String message, Throwable throwable) {
            // No-op
        }
        public void info(Object... messages) {
            // No-op
        }
        public void debug(Object... messages) {
            // No-op
        }
    }

    private Logger logger = new DummyLogger();

    public void keyDown(W3CKeyEvent keyEvent) {
        // No-op
    }

    public void keyUp(W3CKeyEvent keyEvent) {
        // No-op
    }

    public void pointerUp(int button, String sourceId, PointerType pointerType,
                            Long x, Long y, Set<Integer> depressedButtons,
                            KeyInputState globalKeyInputState) throws AppiumException {
        // No-op
    }

    public void pointerDown(int button, String sourceId, PointerType pointerType,
                     Long x, Long y, Set<Integer> depressedButtons,
                     KeyInputState globalKeyInputState) throws AppiumException {
        // No-op
    }

    public void pointerCancel(String sourceId, PointerType pointerType) throws AppiumException {
        // No-op
    }

    public double getPointerMoveDurationMargin(PointerInputState pointerInputState) {
        if (pointerInputState.getType() == PointerType.TOUCH && !pointerInputState.hasPressedButtons()) {
            // If no buttons are pushed nothing happens, so skip to the end
            // of the pointer move
            // e.g.: touch move without pressed buttons is like a finger moving without
            //      being pressed on the screen
            return 1.0;
        }

        // Give a margin of error of 1%
        return 0.01;
    }

    public void pointerMove(String sourceId, PointerType pointerType,
                            long currentX, long currentY, long x, long y,
                            Set<Integer> buttons, KeyInputState globalKeyInputState) throws AppiumException {
        // Add the pointer move event to the logs
        PointerMoveEvent pointerMoveEvent = new PointerMoveEvent();
        pointerMoveEvent.sourceId = sourceId;
        pointerMoveEvent.pointerType = pointerType;
        pointerMoveEvent.currentX = currentX;
        pointerMoveEvent.currentY = currentY;
        pointerMoveEvent.x = x;
        pointerMoveEvent.y = y;
        pointerMoveEvent.buttons = buttons;
        pointerMoveEvent.globalKeyInputState = globalKeyInputState;
        pointerMoveEvents.add(pointerMoveEvent);
    }

    public List<PointerMoveEvent> getPointerMoveEvents() {
        return pointerMoveEvents;
    }

    public Point getElementCenterPoint(String elementId)
            throws NoSuchElementException, StaleElementException, NotYetImplementedException {
        if ("none".equals(elementId)) {
            throw new NoSuchElementException(String.format("Could not find element with id: %s", elementId));
        } else if ("stale".equals(elementId)) {
            throw new StaleElementException(String.format("Element with id %s no longer exists", elementId));
        }

        Point point = new Point();
        point.x = 10;
        point.y = 10;
        return point;
    }

    public long getViewportHeight() {
        return 400;
    }

    public long getViewportWidth() {
        return 200;
    }

    public Logger getLogger() {
        return logger;
    }
}

File: espresso-server/library/src/main/java/io/appium/espressoserver/lib/helpers/w3c/adapter/DummyW3CActionAdapter.kt
package io.appium.espressoserver.lib.helpers.w3c.adapter

import android.graphics.Point

import java.util.ArrayList

import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
import io.appium.espressoserver.lib.handlers.exceptions.NoSuchElementException
import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException
import io.appium.espressoserver.lib.helpers.Logger
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.W3CKeyEvent
import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType
import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState
import io.appium.espressoserver.lib.helpers.w3c.state.PointerInputState

open class DummyW3CActionAdapter : BaseW3CActionAdapter() {

    // Keep a log of pointer move events so the values can be checked in the unit tests
    private val pointerMoveEvents = ArrayList<PointerMoveEvent>()
    override val logger: Logger = DummyLogger()

    override val viewportHeight: Long
        get() = 400

    override val viewportWidth: Long
        get() = 200

    class PointerMoveEvent {
        var sourceId: String? = null
        var pointerType: PointerType? = null
        var currentX: Float = 0.toFloat()
        var currentY: Float = 0.toFloat()
        var x: Float = 0.toFloat()
        var y: Float = 0.toFloat()
        var buttons: Set<Int>? = null
        var globalKeyInputState: KeyInputState? = null
    }

    private inner class DummyLogger : Logger {
        override fun error(vararg messages: Any) {
            // No-op
        }

        override fun error(message: String, throwable: Throwable) {
            // No-op
        }

        override fun info(vararg messages: Any) {
            // No-op
        }

        override fun debug(vararg messages: Any) {
            // No-op
        }

        override fun warn(vararg messages: Any) {
            // No-op
        }
    }

    override fun keyDown(keyDownEvent: W3CKeyEvent) {
        // No-op
    }

    override fun keyUp(keyUpEvent: W3CKeyEvent) {
        // No-op
    }

    @Throws(AppiumException::class)
    override fun pointerUp(button: Int, sourceId: String, pointerType: PointerType?,
                       x: Float, y: Float, depressedButtons: Set<Int>,
                       globalKeyInputState: KeyInputState?) {
        // No-op
    }

    @Throws(AppiumException::class)
    override fun pointerDown(button: Int, sourceId: String, pointerType: PointerType?,
                         x: Float, y: Float, depressedButtons: Set<Int>,
                         globalKeyInputState: KeyInputState?) {
        // No-op
    }

    override fun pointerCancel(sourceId: String, pointerType: PointerType) {
        // No-op
    }

    override fun getPointerMoveDurationMargin(pointerInputState: PointerInputState): Double {
        return if (pointerInputState.type == PointerType.TOUCH && !pointerInputState.hasPressedButtons()) {
            // If no buttons are pushed nothing happens, so skip to the end
            // of the pointer move
            // e.g.: touch move without pressed buttons is like a finger moving without
            //      being pressed on the screen
            1.0
        } else 0.01

        // Give a margin of error of 1%
    }

    override fun pointerMove(sourceId: String, pointerType: PointerType?,
                             currentX: Float, currentY: Float, x: Float, y: Float,
                             buttons: Set<Int>?, globalKeyInputState: KeyInputState?) {
        // Add the pointer move event to the logs
        val pointerMoveEvent = PointerMoveEvent()
        pointerMoveEvent.apply {
            this.sourceId = sourceId
            this.pointerType = pointerType
            this.currentX = currentX
            this.currentY = currentY
            this.x = x
            this.y = y
            this.buttons = buttons
            this.globalKeyInputState = globalKeyInputState
        }
        pointerMoveEvents.add(pointerMoveEvent)
    }

    fun getPointerMoveEvents(): List<PointerMoveEvent> {
        return pointerMoveEvents
    }

    @Throws(AppiumException::class)
    override fun getElementCenterPoint(elementId: String?): Point {
        when(elementId) {
            "none" -> throw NoSuchElementException("Could not find element with id: ${elementId}")
            "stale" -> throw StaleElementException("Element with id ${elementId} no longer exists")
        }

        val point = Point()
        point.x = 10
        point.y = 10
        return point
    }
}
